target_chains/ethereum/contracts/contracts/pulse/Pulse.sol

```diff
@@ -0,0 +1,462 @@
+        require(
+            providerInfo.accruedFeesInWei >= amount,
+            "Insufficient balance"
+        );
+        providerInfo.accruedFeesInWei -= amount;
+
+        // Interaction with an external contract or token transfer
+        (bool sent, ) = msg.sender.call{value: amount}("");
+        require(sent, "withdrawal to msg.sender failed");
+
+        emit ProviderWithdrawn(msg.sender, msg.sender, amount);
+    }
+
+    function withdrawAsFeeManager(
+        address provider,
+        uint128 amount
+    ) external override {
+        ProviderInfo storage providerInfo = _state.providers[provider];
+
+        if (providerInfo.sequenceNumber == 0) {
+            revert NoSuchProvider();
+        }
+
+        if (providerInfo.feeManager != msg.sender) {
+            revert Unauthorized();
+        }
+
+        // Use checks-effects-interactions pattern to prevent reentrancy attacks.
+        require(
+            providerInfo.accruedFeesInWei >= amount,
+            "Insufficient balance"
+        );
+        providerInfo.accruedFeesInWei -= amount;
+
+        // Interaction with an external contract or token transfer
+        (bool sent, ) = msg.sender.call{value: amount}("");
+        require(sent, "withdrawal to msg.sender failed");
+
+        emit ProviderWithdrawn(provider, msg.sender, amount);
+    }
+
+    function requestPriceUpdatesWithCallback(
+        address provider,
+        uint256 publishTime,
+        bytes32[] calldata priceIds,
+        uint256 callbackGasLimit
+    ) external payable override returns (uint64 requestSequenceNumber) {
+        ProviderInfo storage providerInfo = _state.providers[provider];
+        if (providerInfo.sequenceNumber == 0) revert NoSuchProvider();
+
+        if (
+            providerInfo.maxNumPrices > 0 &&
+            priceIds.length > providerInfo.maxNumPrices
+        ) {
+            revert ExceedsMaxPrices(
+                uint32(priceIds.length),
+                providerInfo.maxNumPrices
+            );
+        }
+
+        // Assign sequence number and increment
+        requestSequenceNumber = providerInfo.sequenceNumber++;
+
+        // Verify fee payment
+        uint128 requiredFee = getFee(provider, callbackGasLimit);
+        if (msg.value < requiredFee) revert InsufficientFee();
+
+        // Store request for callback execution
+        Request storage req = allocRequest(provider, requestSequenceNumber);
+        req.provider = provider;
+        req.sequenceNumber = requestSequenceNumber;
+        req.publishTime = publishTime;
+        req.priceIds = priceIds;
+        req.callbackGasLimit = callbackGasLimit;
+        req.requester = msg.sender;
+
+        // Update fee balances
+        providerInfo.accruedFeesInWei += providerInfo.feeInWei;
+        _state.accruedPythFeesInWei +=
+            SafeCast.toUint128(msg.value) -
+            providerInfo.feeInWei;
+
+        emit PriceUpdateRequested(req);
```

Comment:
I think it's good to still emit events so that its easier to track metrics, etc even though argus can just query the storage directly for the request